kratos/processes/calculate_discontinuous_distance_to_skin_process.cpp

```diff
@@ -764,6 +810,140 @@ namespace Kratos
         return is_shared;
     }
 
+    template<std::size_t TDim>
+    void CalculateDiscontinuousDistanceToSkinProcess<TDim>::CheckAndCorrectEdgeDistances()
+    {
+
+        KRATOS_TRY;
+
+        CalculateElementalNeighbours();
+
+        auto p_global_ptr_comm = CreatePointerCommunicator();
+
+        // Proxy to retrieve cut edges from neighbours in other partitions. Works as a lambda in serial runs.
+        auto get_neighbour_cut_edges = p_global_ptr_comm->Apply([](GlobalPointer<Element> pElem) {
+
+            std::unordered_map<IndexType, IndexType> global_id_to_local;
+            for (IndexType i = 0; i < mNumNodes; ++i) {
+                global_id_to_local[pElem->GetGeometry()[i].Id()] = i;
+            }
+
+            auto elem_dist = pElem->GetValue(ELEMENTAL_DISTANCES);
+            std::vector<std::vector<IndexType>> cut_edges_vector;
+            const auto edges_container = pElem->GetGeometry().GenerateEdges();
+            for (std::size_t i_edge = 0; i_edge < mNumEdges; ++i_edge) {
+                IndexType edge_i_id = edges_container[i_edge][0].Id();
+                IndexType edge_j_id = edges_container[i_edge][1].Id();
+                double i_distance = elem_dist[global_id_to_local[edge_i_id]];
+                double j_distance = elem_dist[global_id_to_local[edge_j_id]];
+                bool is_edge_cut = (i_distance * j_distance) < 0.0;
+                std::vector<IndexType> edge_vector;
+                edge_vector.push_back(std::min(edge_i_id, edge_j_id));
+                edge_vector.push_back(std::max(edge_i_id, edge_j_id));
+                if (is_edge_cut) {
+                    cut_edges_vector.push_back(edge_vector);
+                }
+            }
+            return cut_edges_vector;
+        });
+
+        auto check_edge_against_distances = [&, get_neighbour_cut_edges] (const Element::GeometryType::GeometriesArrayType& rEdgeContainer, Vector& rCuteEdgeVector) {
+            for (std::size_t i_edge = 0; i_edge < mNumEdges; ++i_edge) {
+                // Edge initially cut with distances exactly 0
+                if (rCuteEdgeVector[i_edge] >= 0) {
+                    IndexType edge_i_id = std::min(rEdgeContainer[i_edge][0].Id(), rEdgeContainer[i_edge][1].Id());
+                    IndexType edge_j_id = std::max(rEdgeContainer[i_edge][0].Id(), rEdgeContainer[i_edge][1].Id());
+
+                    // Get neighbour elements in a single vector without repetitions.
+                    std::unordered_set<Element::WeakPointer, SharedPointerHasher<Element::WeakPointer>, SharedPointerComparator<Element::WeakPointer>> all_neighbour_set;
+                    for(auto& neigh : rEdgeContainer[i_edge][0].GetValue(NEIGHBOUR_ELEMENTS).GetContainer()) {
+                        all_neighbour_set.insert(neigh);
+                    }
```

Comment:
```suggestion
                    const auto& r_neigh_container = rEdgeContainer[i_edge][0].GetValue(NEIGHBOUR_ELEMENTS).GetContainer();
                    for(const auto& r_neigh : r_neigh_container) {
                        all_neighbour_set.insert(r_neigh);
                    }
```
Otherwise you're accessing the neighbours each time the if condition is evaluated. About the `const`(s), I'm not 100% sure (just give it a try).